### ISSUE TITLE
Registering C-style integer enums?

### ISSUE THREAD

**Issue by Ace4896** [woboq/qmetaobject-rs#39]
How would I register a C-style integer enum e.g.

```rust
enum MyEnum {
    A = 1,
    B = 2,
    // etc
}
```

so that I could use it in both Rust and QML?

Currently, I'm working around this by duplicating the enum definition in my QML and Rust code, where I cast the Rust enum as seen [here](https://stackoverflow.com/a/31359044) and pass around integer values, but is there a way of registering this type in QML directly?

**Comment by ogoffart** [comment]
right, this is not yet implemented.
Not sure how it could work. 
We would need the derive macro to know about the name of each enum value, so that means it probay should be declared inside the struct definition. But rust don't allow that. So it would have to be wrap in a macro or an attribute.

**Comment by Guiguiprim** [comment]
Maybe we could exploit namespace enums (since Qt 5.8 https://www.kdab.com/new-qt-5-8-meta-object-support-namespaces/amp/) to start introducing registered enums.

```rust
#[derive(QEnums)]
enum MyEnum { ... }
```

Could generate a metaobject as if

```cpp
namespace MyEnum {
enum MyEnum { ... };
Q_NAMESPACE
Q_ENUM_NS(MyEnum)
}
```

**Comment by Guiguiprim** [comment]
I'm going to try to add QEnum based on this idea.

I might need to finalize the INT_DATA to create the static QMetaObject during the lazy initialization as I don't think it's possible to get the enums descriminant values during code-gen.

Also I will need to ensure that the property type name match a meta-type name (MyEnum vs MyEnum::MyEnum).

Should be fun!

**Comment by ogoffart** [comment]
Right, so then the enum will be "MyEnum::MyEnum" and there would be one metaobject only for this enum.
I guess that should work.

**Comment by ratijas** [comment]
I tried this code, and it works as of `qmetaobject-rs` v0.1.4:

```rust
use qmetaobject::*;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
#[derive(QEnum)]
#[repr(u32)]
pub enum Role {
    Default = 1,
    Service = 2,
}

let uri = cstr!("MyTest");
qml_register_enum::<Role>(uri, 1, 0, cstr!("Role"));
```

```qml
import MyTest 1.0
Text {
    text: `${Role.Service}`
}
```

It was implemented by @Guiguiprim back in 2019 via #42. Thus, I'm closing this issue. Thanks, Guiguiprim!
